File: src/services/secure-storage.ts
// src/services/secure-storage.ts
import * as keytar from 'keytar';
import * as crypto from 'crypto';
import ElectronStore from 'electron-store';
import { Host } from '../models/host';
import { SSHKey } from '../models/ssh-key';
import { Snippet } from '../models/snippet';
import { UserSettings } from '../models/user-settings';

// Define schema for ElectronStore
const schema = {
  hosts: {
    type: 'object' as const,
    additionalProperties: {
      type: 'object' as const,
      properties: {
        // Host properties without sensitive data
      },
    },
  },
  sshKeys: {
    type: 'object' as const,
    additionalProperties: {
      type: 'object' as const,
      properties: {
        // SSH key properties without sensitive data
      },
    },
  },
  snippets: {
    type: 'object' as const,
    additionalProperties: {
      type: 'object' as const,
      properties: {
        // Snippet properties
      },
    },
  },
  settings: {
    type: 'object' as const,
    properties: {
      // User settings properties
    },
  },
  hostGroups: {
    type: 'object' as const,
    additionalProperties: {
      type: 'object' as const,
      properties: {
        // Host group properties
      },
    },
  },
};

// Define the store data structure type
type StoreSchema = {
  hosts: Record<string, any>;
  sshKeys: Record<string, any>;
  snippets: Record<string, any>;
  settings: any;
  hostGroups: Record<string, any>;
};

export class SecureStorage {
  private store: ElectronStore<StoreSchema>;
  private encryptionKey: Buffer | null = null;
  private readonly SERVICE_NAME = 'Teemux';
  private readonly ENCRYPTION_KEY_ACCOUNT = 'encryption-key';

  constructor() {
    this.store = new ElectronStore<StoreSchema>({ schema });
    this.initializeEncryptionKey();
  }

  private async initializeEncryptionKey(): Promise<void> {
    try {
      // Try to retrieve existing encryption key
      let key = await keytar.getPassword(
        this.SERVICE_NAME,
        this.ENCRYPTION_KEY_ACCOUNT
      );

      if (!key) {
        // Generate a new encryption key if none exists
        key = crypto.randomBytes(32).toString('hex');
        await keytar.setPassword(
          this.SERVICE_NAME,
          this.ENCRYPTION_KEY_ACCOUNT,
          key
        );
      }

      this.encryptionKey = Buffer.from(key, 'hex');
    } catch (error) {
      console.error('Failed to initialize encryption key:', error);
      throw new Error('Failed to initialize secure storage');
    }
  }

  private encrypt(data: string): string {
    if (!this.encryptionKey) {
      throw new Error('Encryption key not initialized');
    }

    const iv = crypto.randomBytes(16);
    const cipher = crypto.createCipheriv('aes-256-gcm', this.encryptionKey, iv);

    let encrypted = cipher.update(data, 'utf8', 'hex');
    encrypted += cipher.final('hex');

    const authTag = cipher.getAuthTag();

    // Return IV + Auth Tag + Encrypted Data
    return iv.toString('hex') + ':' + authTag.toString('hex') + ':' + encrypted;
  }

  private decrypt(data: string): string {
    if (!this.encryptionKey) {
      throw new Error('Encryption key not initialized');
    }

    const parts = data.split(':');
    if (parts.length !== 3) {
      throw new Error('Invalid encrypted data format');
    }

    const iv = Buffer.from(parts[0], 'hex');
    const authTag = Buffer.from(parts[1], 'hex');
    const encryptedText = parts[2];

    const decipher = crypto.createDecipheriv(
      'aes-256-gcm',
      this.encryptionKey,
      iv
    );
    decipher.setAuthTag(authTag);

    let decrypted = decipher.update(encryptedText, 'hex', 'utf8');
    decrypted += decipher.final('utf8');

    return decrypted;
  }

  // Host management
  async saveHost(host: Host): Promise<void> {
    // Create a copy without sensitive data for the store
    const hostForStore = { ...host };

    // Store sensitive data securely
    if (host.password) {
      await keytar.setPassword(
        this.SERVICE_NAME,
        `host-${host.id}-password`,
        host.password
      );
      delete hostForStore.password;
    }

    if (host.privateKeyPassphrase) {
      await keytar.setPassword(
        this.SERVICE_NAME,
        `host-${host.id}-key-passphrase`,
        host.privateKeyPassphrase
      );
      delete hostForStore.privateKeyPassphrase;
    }

    // Save non-sensitive data to store
    const hosts = this.store.get('hosts', {}) as Record<string, any>;
    hosts[host.id] = hostForStore;
    this.store.set('hosts', hosts);
  }

  async getHost(hostId: string): Promise<Host | null> {
    const hosts = this.store.get('hosts', {}) as Record<string, any>;
    const host = hosts[hostId];

    if (!host) {
      return null;
    }

    // Retrieve sensitive data
    try {
      const password = await keytar.getPassword(
        this.SERVICE_NAME,
        `host-${hostId}-password`
      );
      const keyPassphrase = await keytar.getPassword(
        this.SERVICE_NAME,
        `host-${hostId}-key-passphrase`
      );

      return {
        ...host,
        password: password || undefined,
        privateKeyPassphrase: keyPassphrase || undefined,
      };
    } catch (error) {
      console.error('Error retrieving sensitive host data:', error);
      return host;
    }
  }

  async getAllHosts(): Promise<Host[]> {
    const hosts = this.store.get('hosts', {}) as Record<string, any>;
    return Promise.all(
      Object.keys(hosts).map((hostId) => this.getHost(hostId))
    ).then((results) => results.filter(Boolean) as Host[]);
  }

  async deleteHost(hostId: string): Promise<void> {
    // Delete sensitive data
    await keytar.deletePassword(this.SERVICE_NAME, `host-${hostId}-password`);
    await keytar.deletePassword(
      this.SERVICE_NAME,
      `host-${hostId}-key-passphrase`
    );

    // Delete from store
    const hosts = this.store.get('hosts', {}) as Record<string, any>;
    delete hosts[hostId];
    this.store.set('hosts', hosts);
  }

  // SSH Key management
  async saveSSHKey(key: SSHKey): Promise<void> {
    // Create a copy without sensitive data for the store
    const keyForStore = { ...key };

    // Store passphrase securely if it exists
    if (key.passphrase) {
      await keytar.setPassword(
        this.SERVICE_NAME,
        `key-${key.id}-passphrase`,
        key.passphrase
      );
      delete keyForStore.passphrase;
    }

    // Save non-sensitive data to store
    const keys = this.store.get('sshKeys', {}) as Record<string, any>;
    keys[key.id] = keyForStore;
    this.store.set('sshKeys', keys);
  }

  async getSSHKey(keyId: string): Promise<SSHKey | null> {
    const keys = this.store.get('sshKeys', {}) as Record<string, any>;
    const key = keys[keyId];

    if (!key) {
      return null;
    }

    // Retrieve passphrase if it exists
    try {
      const passphrase = await keytar.getPassword(
        this.SERVICE_NAME,
        `key-${keyId}-passphrase`
      );

      return {
        ...key,
        passphrase: passphrase || undefined,
      };
    } catch (error) {
      console.error('Error retrieving SSH key passphrase:', error);
      return key;
    }
  }

  async getAllSSHKeys(): Promise<SSHKey[]> {
    const keys = this.store.get('sshKeys', {}) as Record<string, any>;
    return Promise.all(
      Object.keys(keys).map((keyId) => this.getSSHKey(keyId))
    ).then((results) => results.filter(Boolean) as SSHKey[]);
  }

  async deleteSSHKey(keyId: string): Promise<void> {
    // Delete passphrase if it exists
    await keytar.deletePassword(this.SERVICE_NAME, `key-${keyId}-passphrase`);

    // Delete from store
    const keys = this.store.get('sshKeys', {}) as Record<string, any>;
    delete keys[keyId];
    this.store.set('sshKeys', keys);
  }

  // Snippet management
  async saveSnippet(snippet: Snippet): Promise<void> {
    const snippets = this.store.get('snippets', {}) as Record<string, any>;
    snippets[snippet.id] = snippet;
    this.store.set('snippets', snippets);
  }

  async getSnippet(snippetId: string): Promise<Snippet | null> {
    const snippets = this.store.get('snippets', {}) as Record<string, any>;
    return snippets[snippetId] || null;
  }

  async getAllSnippets(): Promise<Snippet[]> {
    const snippets = this.store.get('snippets', {}) as Record<string, any>;
    return Object.values(snippets);
  }

  async deleteSnippet(snippetId: string): Promise<void> {
    const snippets = this.store.get('snippets', {}) as Record<string, any>;
    delete snippets[snippetId];
    this.store.set('snippets', snippets);
  }

  // User settings
  async saveUserSettings(settings: UserSettings): Promise<void> {
    this.store.set('settings', settings);
  }

  async getUserSettings(): Promise<UserSettings> {
    return this.store.get('settings') as UserSettings;
  }

  // Cloud sync methods
  async exportEncryptedData(masterPassword: string): Promise<string> {
    // Get all data
    const hosts = await this.getAllHosts();
    const sshKeys = await this.getAllSSHKeys();
    const snippets = await this.getAllSnippets();
    const settings = await this.getUserSettings();

    // Create export object
    const exportData = {
      hosts,
      sshKeys,
      snippets,
      settings,
      exportDate: new Date(),
    };

    // Encrypt with master password
    const salt = crypto.randomBytes(16);
    const key = crypto.pbkdf2Sync(masterPassword, salt, 100000, 32, 'sha256');

    const iv = crypto.randomBytes(16);
    const cipher = crypto.createCipheriv('aes-256-gcm', key, iv);

    let encrypted = cipher.update(JSON.stringify(exportData), 'utf8', 'hex');
    encrypted += cipher.final('hex');

    const authTag = cipher.getAuthTag();

    // Return salt + IV + Auth Tag + Encrypted Data
    return (
      salt.toString('hex') +
      ':' +
      iv.toString('hex') +
      ':' +
      authTag.toString('hex') +
      ':' +
      encrypted
    );
  }

  async importEncryptedData(
    encryptedData: string,
    masterPassword: string
  ): Promise<boolean> {
    try {
      const parts = encryptedData.split(':');
      if (parts.length !== 4) {
        throw new Error('Invalid encrypted data format');
      }

      const salt = Buffer.from(parts[0], 'hex');
      const iv = Buffer.from(parts[1], 'hex');
      const authTag = Buffer.from(parts[2], 'hex');
      const encryptedText = parts[3];

      const key = crypto.pbkdf2Sync(masterPassword, salt, 100000, 32, 'sha256');

      const decipher = crypto.createDecipheriv('aes-256-gcm', key, iv);
      decipher.setAuthTag(authTag);

      let decrypted = decipher.update(encryptedText, 'hex', 'utf8');
      decrypted += decipher.final('utf8');

      const importData = JSON.parse(decrypted);

      // Import hosts
      for (const host of importData.hosts) {
        await this.saveHost(host);
      }

      // Import SSH keys
      for (const key of importData.sshKeys) {
        await this.saveSSHKey(key);
      }

      // Import snippets
      for (const snippet of importData.snippets) {
        await this.saveSnippet(snippet);
      }

      // Import settings
      if (importData.settings) {
        await this.saveUserSettings(importData.settings);
      }

      return true;
    } catch (error) {
      console.error('Import failed:', error);
      return false;
    }
  }

  // Pro feature check
  async isProUser(): Promise<boolean> {
    const settings = await this.getUserSettings();
    return settings?.proUser || false;
  }

  // Feature gating
  async canUseFeature(featureName: string): Promise<boolean> {
    // List of features that require Pro subscription
    const proFeatures = [
      'cloud-sync',
      'sftp',
      'port-forwarding',
      'advanced-snippets',
      'jump-host',
      'multi-factor-auth',
    ];

    // If it's not a pro feature, allow it
    if (!proFeatures.includes(featureName)) {
      return true;
    }

    // Check if user has Pro subscription
    return this.isProUser();
  }

  // Host group methods
  async saveHostGroup(group: any): Promise<void> {
    const hostGroups = this.store.get('hostGroups') || {};
    hostGroups[group.id] = group;
    this.store.set('hostGroups', hostGroups);
  }

  async getHostGroup(groupId: string): Promise<any | null> {
    const hostGroups = this.store.get('hostGroups') || {};
    return hostGroups[groupId] || null;
  }

  async getAllHostGroups(): Promise<any[]> {
    const hostGroups = this.store.get('hostGroups') || {};
    return Object.values(hostGroups);
  }

  async deleteHostGroup(groupId: string): Promise<void> {
    const hostGroups = this.store.get('hostGroups') || {};
    delete hostGroups[groupId];
    this.store.set('hostGroups', hostGroups);
  }
}

File: src-tauri/src/services/secure_storage.rs
use aes_gcm::{
  aead::{Aead, KeyInit},
  Aes256Gcm, Nonce,
};
use anyhow::{Context, Result};
use base64::{engine::general_purpose::STANDARD as BASE64, Engine as _};
use chrono::{DateTime, Utc};
use keyring::Entry;
use rand::RngCore;
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::{collections::HashMap, fs, path::PathBuf};
use tokio::sync::Mutex;
// Import models
use crate::models::{
  host::{Host, HostGroup},
  snippet::Snippet,
  ssh_key::SSHKey,
  user_settings::UserSettings,
};

#[derive(Debug, Clone, Serialize, Deserialize)]
struct Store {
  hosts: HashMap<String, Host>,
  snippets: HashMap<String, Snippet>,
  settings: Option<UserSettings>,
  host_groups: HashMap<String, HostGroup>,
  ssh_keys: HashMap<String, SSHKey>,
}

impl Store {
  fn load() -> Result<Self> {
    let store_path = Self::get_store_path()?;
    if store_path.exists() {
      let data = fs::read_to_string(&store_path)?;
      Ok(serde_json::from_str(&data)?)
    } else {
      Ok(Self {
        hosts: HashMap::new(),
        snippets: HashMap::new(),
        settings: None,
        host_groups: HashMap::new(),
        ssh_keys: HashMap::new(),
      })
    }
  }

  fn save(&self) -> Result<()> {
    let store_path = Self::get_store_path()?;
    let data = serde_json::to_string_pretty(self)?;
    fs::write(&store_path, data)?;
    Ok(())
  }

  fn get_store_path() -> Result<PathBuf> {
    let mut path = dirs::config_dir()
      .context("Failed to get config directory")?
      .join("teemux");
    fs::create_dir_all(&path)?;
    path.push("store.json");
    Ok(path)
  }
}

pub struct SecureStorage {
  store: Mutex<Store>,
  encryption_key: Vec<u8>,
  service_name: String,
}

impl SecureStorage {
  pub async fn new() -> Result<Self> {
    let encryption_key = Self::initialize_encryption_key().await?;
    let store = Store::load()?;

    Ok(Self {
      store: Mutex::new(store),
      encryption_key,
      service_name: "Teemux".to_string(),
    })
  }

  async fn initialize_encryption_key() -> Result<Vec<u8>> {
    let keyring = Entry::new("Teemux", "encryption-key")?;

    match keyring.get_password() {
      Ok(key) => Ok(hex::decode(key)?),
      Err(_) => {
        let new_key = rand::random::<[u8; 32]>();
        let hex_key = hex::encode(&new_key);
        keyring.set_password(&hex_key)?;
        Ok(new_key.to_vec())
      }
    }
  }

  fn encrypt(&self, data: &str) -> Result<String> {
    let cipher = Aes256Gcm::new_from_slice(&self.encryption_key)
      .map_err(|e| anyhow::anyhow!("Failed to create cipher: {}", e))?;
    let nonce = Nonce::from_slice(&self.encryption_key[..12]);
    let ciphertext = cipher
      .encrypt(nonce, data.as_bytes())
      .map_err(|e| anyhow::anyhow!("Failed to encrypt data: {}", e))?;
    Ok(BASE64.encode(ciphertext))
  }

  fn decrypt(&self, data: &str) -> Result<String> {
    let cipher = Aes256Gcm::new_from_slice(&self.encryption_key)
      .map_err(|e| anyhow::anyhow!("Failed to create cipher: {}", e))?;
    let nonce = Nonce::from_slice(&self.encryption_key[..12]);
    let ciphertext = BASE64.decode(data)?;
    let decrypted = cipher
      .decrypt(nonce, ciphertext.as_slice())
      .map_err(|e| anyhow::anyhow!("Failed to decrypt data: {}", e))?;
    Ok(String::from_utf8(decrypted)?)
  }

  pub async fn save_host(&self, host: Host) -> Result<()> {
    let mut store = self.store.lock().await;

    // Handle sensitive data
    if let Some(password) = &host.password {
      let entry = Entry::new(&self.service_name, &format!("host-{}-password", host.id))?;
      entry.set_password(&self.encrypt(password)?)?;
    }

    if let Some(passphrase) = &host.private_key_passphrase {
      let entry = Entry::new(
        &self.service_name,
        &format!("host-{}-key-passphrase", host.id),
      )?;
      entry.set_password(&self.encrypt(passphrase)?)?;
    }

    // Store non-sensitive data
    let host_data = host.clone();
    store.hosts.insert(host.id.clone(), host_data);
    store.save()?;

    Ok(())
  }

  pub async fn get_host(&self, host_id: &str) -> Result<Option<Host>> {
    let store = self.store.lock().await;

    if let Some(host) = store.hosts.get(host_id) {
      let mut host = host.clone();

      // Retrieve sensitive data
      let password_entry = Entry::new(&self.service_name, &format!("host-{}-password", host_id))?;

      if let Ok(encrypted_password) = password_entry.get_password() {
        host.password = Some(self.decrypt(&encrypted_password)?);
      }

      let passphrase_entry = Entry::new(
        &self.service_name,
        &format!("host-{}-key-passphrase", host_id),
      )?;

      if let Ok(encrypted_passphrase) = passphrase_entry.get_password() {
        host.private_key_passphrase = Some(self.decrypt(&encrypted_passphrase)?);
      }

      Ok(Some(host))
    } else {
      Ok(None)
    }
  }

  pub async fn get_all_hosts(&self) -> Result<Vec<Host>> {
    let store = self.store.lock().await;
    Ok(store.hosts.values().cloned().collect())
  }

  pub async fn delete_host(&self, host_id: &str) -> Result<()> {
    let mut store = self.store.lock().await;
    store.hosts.remove(host_id);
    store.save()?;
    Ok(())
  }

  // ... similar implementations for SSH keys, snippets, etc.
  pub async fn save_ssh_key(&self, key: SSHKey) -> Result<()> {
    let mut store = self.store.lock().await;
    store.ssh_keys.insert(key.id.clone(), key);

    store.save()?;
    Ok(())
  }

  pub async fn get_ssh_key(&self, key_id: &str) -> Result<Option<SSHKey>> {
    let store = self.store.lock().await;
    Ok(store.ssh_keys.get(key_id).cloned())
  }

  pub async fn get_all_ssh_keys(&self) -> Result<Vec<SSHKey>> {
    let store = self.store.lock().await;
    Ok(store.ssh_keys.values().cloned().collect())
  }

  pub async fn delete_ssh_key(&self, key_id: &str) -> Result<()> {
    let mut store = self.store.lock().await;
    store.ssh_keys.remove(key_id);
    store.save()?;

    // todo remve all host using this ssh key
    Ok(())
  }

  // Snippet management
  pub async fn save_snippet(&self, snippet: Snippet) -> Result<()> {
    let mut store = self.store.lock().await;
    store.snippets.insert(snippet.id.clone(), snippet);
    store.save()?;
    Ok(())
  }

  pub async fn get_snippet(&self, snippet_id: &str) -> Result<Option<Snippet>> {
    let store = self.store.lock().await;
    Ok(store.snippets.get(snippet_id).cloned())
  }

  pub async fn get_all_snippets(&self) -> Result<Vec<Snippet>> {
    let store = self.store.lock().await;
    Ok(store.snippets.values().cloned().collect())
  }

  pub async fn delete_snippet(&self, snippet_id: &str) -> Result<()> {
    let mut store = self.store.lock().await;
    store.snippets.remove(snippet_id);
    store.save()?;
    Ok(())
  }

  // User settings
  pub async fn save_user_settings(&self, settings: UserSettings) -> Result<()> {
    let mut store = self.store.lock().await;
    store.settings = Some(settings);
    store.save()?;
    Ok(())
  }

  pub async fn get_user_settings(&self) -> Result<Option<UserSettings>> {
    let store = self.store.lock().await;
    Ok(store.settings.clone())
  }

  // Cloud sync
  pub async fn export_encrypted_data(&self, master_password: &str) -> Result<String> {
    let store = self.store.lock().await;

    // Create export data
    let export_data = serde_json::json!({
        "hosts": store.hosts,
        "snippets": store.snippets,
        "settings": store.settings,
        "host_groups": store.host_groups,
        "export_date": Utc::now()
    });

    // Generate salt and key
    let mut salt = [0u8; 16];
    rand::rng().fill_bytes(&mut salt);

    // Derive key
    let mut hasher = Sha256::default();
    hasher.update(master_password.as_bytes());
    hasher.update(salt);
    let key = hasher.finalize();

    // Encrypt data
    let cipher = Aes256Gcm::new_from_slice(&key)?;
    let nonce = Nonce::from_slice(&salt[..12]); // Use part of salt as nonce

    let data = serde_json::to_string(&export_data)?;
    let encrypted = cipher
      .encrypt(nonce, data.as_bytes())
      .map_err(|e| anyhow::anyhow!("Failed to encrypt data: {}", e))?;

    // Combine salt and encrypted data
    let mut result = Vec::with_capacity(salt.len() + encrypted.len());
    result.extend_from_slice(&salt);
    result.extend_from_slice(&encrypted);

    Ok(BASE64.encode(result))
  }

  pub async fn import_encrypted_data(
    &self,
    encrypted_data: &str,
    master_password: &str,
  ) -> Result<bool> {
    let data = BASE64.decode(encrypted_data)?;
    if data.len() < 16 {
      return Ok(false);
    }

    // Extract salt and encrypted data
    let (salt, encrypted) = data.split_at(16);

    // Derive key
    let mut hasher = Sha256::default();
    hasher.update(master_password.as_bytes());
    hasher.update(salt);
    let key = hasher.finalize();

    // Decrypt data
    let cipher = Aes256Gcm::new_from_slice(&key)?;
    let nonce = Nonce::from_slice(&salt[..12]);

    let decrypted = match cipher.decrypt(nonce, encrypted.as_ref()) {
      Ok(data) => data,
      Err(_) => return Ok(false),
    };
    let decrypted_str = String::from_utf8(decrypted)?;
    let import_data: serde_json::Value = serde_json::from_str(&decrypted_str)?;

    // Import data
    let mut store = self.store.lock().await;

    if let Some(hosts) = import_data["hosts"].as_object() {
      store.hosts = serde_json::from_value(import_data["hosts"].clone())?;
    }

    if let Some(snippets) = import_data["snippets"].as_object() {
      store.snippets = serde_json::from_value(import_data["snippets"].clone())?;
    }

    if let Some(settings) = import_data["settings"].as_object() {
      store.settings = Some(serde_json::from_value(import_data["settings"].clone())?);
    }

    if let Some(host_groups) = import_data["host_groups"].as_object() {
      store.host_groups = serde_json::from_value(import_data["host_groups"].clone())?;
    }

    store.save()?;
    Ok(true)
  }

  // Pro feature check
  pub async fn is_pro_user(&self) -> Result<bool> {
    let store = self.store.lock().await;
    Ok(store.settings.as_ref().map_or(false, |s| s.pro_user))
  }

  pub async fn can_use_feature(&self, feature_name: &str) -> Result<bool> {
    let pro_features = [
      "cloud-sync",
      "sftp",
      "port-forwarding",
      "advanced-snippets",
      "jump-host",
      "multi-factor-auth",
    ];

    if !pro_features.contains(&feature_name) {
      return Ok(true);
    }

    self.is_pro_user().await
  }

  // Host group methods
  pub async fn save_host_group(&self, group: HostGroup) -> Result<()> {
    let mut store = self.store.lock().await;
    store.host_groups.insert(group.id.clone(), group);
    store.save()?;
    Ok(())
  }

  pub async fn get_host_group(&self, group_id: &str) -> Result<Option<HostGroup>> {
    let store = self.store.lock().await;
    Ok(store.host_groups.get(group_id).cloned())
  }

  pub async fn get_all_host_groups(&self) -> Result<Vec<HostGroup>> {
    let store = self.store.lock().await;
    Ok(store.host_groups.values().cloned().collect())
  }

  pub async fn delete_host_group(&self, group_id: &str) -> Result<()> {
    let mut store = self.store.lock().await;
    store.host_groups.remove(group_id);
    store.save()?;
    Ok(())
  }
}
